source/slang/syntax.cpp

```diff
@@ -940,6 +940,155 @@ void Type::accept(IValVisitor* visitor, void* extra)
+        if (auto parentDeclRefType = sourceType->As<DeclRefType>())
+        {
+            auto parentDeclRef = parentDeclRefType->declRef;
+            DeclRef<AggTypeDecl> newParentDeclRef = parentDeclRef.As<AggTypeDecl>();
+            // search for a substitution that might apply to us
+            for (auto s = subst; s; s = s->outer.Ptr())
+            {
+                // the generic decl associated with the substitution list must be
+                // the generic decl that declared this parameter
+                auto genericDecl = s->genericDecl;
+                if (genericDecl != parentDeclRef.getDecl()->ParentDecl)
+                    continue;
+                int index = 0;
+                for (auto m : genericDecl->Members)
+                {
+                    if (m.Ptr() == parentDeclRef.getDecl())
+                    {
+                        // We've found it, so return the corresponding specialization argument
+                        (*ioDiff)++;
+                        if (auto declRef = s->args[index].As<DeclRefType>())
+                        {
+                            newParentDeclRef = (*declRef).declRef.As<AggTypeDecl>();
+                            goto searchEnd;
+                        }
+                    }
+                    else if (auto typeParam = m.As<GenericTypeParamDecl>())
+                    {
+                        index++;
+                    }
+                    else if (auto valParam = m.As<GenericValueParamDecl>())
+                    {
+                        index++;
+                    }
+                    else
+                    {
+                    }
+                }
+            }
+        searchEnd:
+            if (newParentDeclRef)
+            {
+                Decl* targetTypeDecl = nullptr;
+                if (newParentDeclRef.getDecl()->memberDictionary.TryGetValue(this->GetDeclRef().decl->getName(), targetTypeDecl))
+                {
+                    if (auto typeDefDecl = targetTypeDecl->As<TypeDefDecl>())
+                        return GetType(DeclRef<TypeDefDecl>(typeDefDecl, subst));
+                    else
+                        return DeclRefType::Create(this->getSession(), DeclRef<Decl>(targetTypeDecl, subst));
+                }
+            }
+        }
+       
+        return this;
+    }
+
+    int AssocTypeDeclRefType::GetHashCode()
+    {
+        return declRef.GetHashCode();
+    }
+
+    Type* AssocTypeDeclRefType::CreateCanonicalType()
+    {
+        return this;
+    }
+
+    // GenericConstraintDeclRefType
+
+    String GenericConstraintDeclRefType::ToString()
+    {
+        // TODO: what is appropriate here?
+        return "<GenericConstraintType>";
+    }
+
+    bool GenericConstraintDeclRefType::EqualsImpl(Type * type)
+    {
+        if (auto other = type->As<GenericConstraintDeclRefType>())
+        {
+            return supType->Equals(other->supType) && subType->Equals(other->subType);
```

Comment:
I'm so confused about what this class of type is supposed to represent. This seems to be a type of the form `Sub : Sup` and we are saying two of them are equal if theirs `Sub`s and `Sup`s are equal... But what does this represent as a *type*?

This actually seems like it may just be a duplicate of the existing `SubtypeWitness` class, but in the `Type` hierarchy instead of the `Witness` hierarchy. If that is the case, and we really need these to be able to appear as, e.g., the `type` of an expression, then we should just make `Witness` inherit from `Type` (since `Type` is representing classifiers at all meta-levels given the existence of `TypeType`).